pywhy_stats/kernel_utils.py

```diff
@@ -0,0 +1,281 @@
+
+    Parameters
+    ----------
+    X : ArrayLike
+        Data array. If the data type is not string, the RBF kernel is used.
+        If data type is string, the delta kernel is used.
+
+    Returns
+    -------
+    Callable[[ArrayLike], ArrayLike]
+        The kernel function.
+    """
+    if X.dtype.type not in (np.str_, np.object_):
+        return "rbf", {"gamma": 0.5 * estimate_squared_sigma_rbf(X)}
+    else:
+        return "delta", dict()
+
+
+def _normalize_data(X: ArrayLike) -> ArrayLike:
+    """Normalize data to zero mean and unit variance."""
+    for column in range(X.shape[1]):
+        if isinstance(X[0, column], int) or isinstance(X[0, column], float):
+            X[:, column] = stats.zscore(X[:, column])
+            X[:, column] = np.nan_to_num(X[:, column])  # in case some dim of X is constant
+    return X
+
+
+def compute_kernel(
+    X: ArrayLike,
+    Y: Optional[ArrayLike] = None,
+    metric: Optional[Union[Callable, str]] = None,
+    centered: bool = True,
+    n_jobs: Optional[int] = None,
+) -> Tuple[ArrayLike, float]:
+    """Compute a kernel matrix and corresponding width.
+
+    Also optionally estimates the kernel width parameter.
+
+    Parameters
+    ----------
+    X : ArrayLike of shape (n_samples_X, n_features_X)
+        The X array.
+    Y : ArrayLike of shape (n_samples_Y, n_features_Y), optional
+        The Y array, by default None.
+    metric : str, optional
+        The metric to compute the kernel function, by default 'rbf'.
+        Can be any string as defined in
+        :func:`sklearn.metrics.pairwise.pairwise_kernels`. Note 'rbf'
+        and 'gaussian' are the same metric.
+    centered : bool, optional
+        Whether to center the kernel matrix or not, by default True.
+        When centered, the kernel matrix induces a zero mean. The main purpose of
+        centering is to remove the bias or mean shift from the data represented
+        by the kernel matrix.
+    n_jobs : int, optional
+        The number of jobs to run computations in parallel, by default None.
+
+    Returns
+    -------
+    kernel : ArrayLike of shape (n_samples_X, n_samples_X) or (n_samples_X, n_samples_Y)
+        The kernel matrix.
+    """
+    # if the width of the kernel is not set, then use the median trick to set the
+    # kernel width based on the data X
+    if metric is None:
+        metric, kernel_params = _get_default_kernel(X)
+    else:
+        kernel_params = dict()
+
+    # compute the potentially pairwise kernel matrix
+    # If the number of arguments is just one, then we bypass the pairwise kernel
+    # optimized computation via sklearn and opt to use the metric function directly
+    if callable(metric) and len(inspect.getfullargspec(metric).args) == 1:
+        kernel = metric(X)
+    else:
+        kernel = pairwise_kernels(
+            X, Y=Y, metric=metric, n_jobs=n_jobs, filter_params=False, **kernel_params
+        )
```

Comment:
Can we move the change of allowing strings as kernels instead of only callable into a separate PR? Would like to take a separate look at this, which is not mixed together with other changes. For this PR, maybe stick to passing a callable (e.g., a kernel from scikit-learn via `kernel_X = partial(pairwise_kernels, metric=metric, n_jobs=n_jobs, filter_params=False, other_params...)` which should be equivalent).

Reply:
Sure thing!

Note: The change here is made to allow maximum flexibility on the user, while not having to write our own pairwise kernel code for custom kernels.

tldr: if we adhere to the scikit-learn pattern, we have less code and buys some helper functions for free, but we need to restrict what type of kernel functions are allowed. 

The issue here is that custom kernels are slow, but that is always the case. However, in order to provide some consistency to make maintenance and code simpler, we probably want to be consistent with how we define a kernel function. That is why I added the extra unused parameters to the delta kernel. Having the `def kernel_func(X, Y=None, **kwargs)` pattern, allows us to re-use the `pairwise_kernels` function.

But it also buys us the usages of strings for free, since they accept those.

Alternatively, if we want to not allow strings, and use the `partial(func(...))` pattern, then we have to write our own parallelized pairwise_kernel function.

Limiting the scope of what we support is probably a good idea, so users can't just feed in some arbitrary lambda functions that we can't easily error check. 

Idk how to handle the `custom_kernel` you define that combines a delta and rbf kernel, so I added this hack for now. However, this raises a question for me: is a custom kernel like that even valid?

Reply:
I think most users can simply use a kernel from scikit. We can definitely think about allowing strings as well, just want to have this review separate.

> However, this raises a question for me: is a custom kernel like that even valid

It is, a convolution of different valid kernels is also a valid kernel and, practically, you can achieve this by pointwise multiplication of the resulting kernel matrices.